cellxgene_resources/curation_qa.ipynb

```diff
@@ -400,46 +400,16 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "def TENx_barcode_checker(df, num_to_check):\n",
-    "    if df.shape[0] < num_to_check:\n",
-    "        num_to_check = df.shape[0]\n",
-    "\n",
-    "    ref_dir = 'ref_files/'\n",
-    "    v2_file = '737K-august-2016.txt'\n",
-    "    v3_file = '3M-february-2018.txt'\n",
-    "    multiome_file = '737K-arc-v1.txt'\n",
-    "\n",
-    "    v2_list = [line.strip() for line in open(ref_dir + v2_file, 'r')]\n",
-    "    v3_list = [line.strip() for line in open(ref_dir + v3_file, 'r')]\n",
-    "    multiome_list = [line.strip() for line in open(ref_dir + multiome_file, 'r')]\n",
-    "\n",
-    "    cellcount = df.index.shape[0]\n",
-    "    barcode_pattern = '[ACTG]{16}'\n",
-    "    barcode_results = ''\n",
-    "    if re.search(barcode_pattern, df.index[0]):\n",
-    "        cellcount\n",
-    "        random_indices = [randint(0, cellcount - 1) for p in range(0, num_to_check)]\n",
-    "        barcodes = {'3pv2_5pv1_5pv2': 0,'3pv3': 0,'multiome': 0,'multiple': 0,'none': 0}\n",
-    "        for i in random_indices:\n",
-    "            if re.search(barcode_pattern, df.index[i]):\n",
-    "                barcode = re.search(barcode_pattern, df.index[i]).group(0)\n",
-    "                if barcode in v2_list:\n",
-    "                    if barcode in multiome_list:\n",
-    "                        barcodes['multiple'] += 1\n",
-    "                    elif barcode in v3_list:\n",
-    "                        barcodes['multiple'] += 1\n",
-    "                    else:\n",
-    "                        barcodes['3pv2_5pv1_5pv2'] += 1\n",
-    "                elif barcode in multiome_list:\n",
-    "                    if barcode in v3_list:\n",
-    "                        barcodes['multiple'] += 1\n",
-    "                    else:\n",
-    "                        barcodes['multiome'] += 1\n",
-    "                elif barcode in v3_list:\n",
-    "                    barcodes['3pv3'] += 1\n",
-    "                else:\n",
-    "                    barcodes['none'] += 1\n",
-    "        return barcodes"
+    "def TENx_barcode_checker(ref_df, obs_df, num_to_check):\n",
+    "    obs_df_sample = obs_df.sample(num_to_check, axis=0)  # can add random_state=1 for reproducibility \n",
+    "    obs_df_split = obs_df_sample.index.str.split('([ACTG]{16})')\n",
+    "    barcodes = pd.DataFrame([b for l in obs_df_split for b in l if re.match(r\".*[ACTG]{16}.*\", b)])\n",
+    "    barcodes.rename(columns={0:'barcode'},inplace=True)\n",
```

Comment:
To handle subsets of data that don't have any barcode to check, you'll need to add a check
after defining `barcodes`, add these 2 lines
```
if barcodes.empty:
    return pd.DataFrame({'summary':['no barcode'] * num_to_check})
```
Should add a row to the report table like the others, but will just report how many observations were checked for barcodes without finding one

Reply:
I found a hiccup in the code when generating `num_to_check`. Not sure how likely this will happen, but for `num_to_check = obs_df.shape[0] // number` where the number used in floor division results in `num_to_check =0`. Where there are low numbers of barcodes associated with a particular donor/sample, but still barcodes we can check?

Reply:
Ya, we should still check the low numbers
Although I think it's unlikely to happen often
We can always update later based on behavior of common use

Reply:
i'll make a note